### PR TITLE
bf: CLDSRV-12 bucket policy resource security fix

### PR DESCRIPTION
**Pull request by naren-rajendran**
Adds resource check for bucket policies. S3 actions are modified to pass request object for resource validation.

### CODE REVIEW

File: lib/api/apiUtils/authorization/permissionChecks.js [hunk 1]
```diff
@@ -182,6 +182,17 @@ function _checkBucketPolicyActions(requestType, actions, log) {
     return evaluators.isActionApplicable(mappedAction, actions, log);
 }
 
+function _checkBucketPolicyResources(request, resource, log) {
+    if (!request || (Array.isArray(resource) && resource.length === 0)) {
```
Comment (miniscruff) [inline lib/api/apiUtils/authorization/permissionChecks.js:186]: Can skip the parenthesis, does not matter much though.
```suggestion
    if (!request || !Array.isArray(resource) || resource.length === 0)
```

File: lib/api/apiUtils/authorization/permissionChecks.js [hunk 2]
```diff
@@ -238,20 +249,21 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
         const s = copiedStatement[0];
         const principalMatch = _checkPrincipals(canonicalID, arn, s.Principal);
         const actionMatch = _checkBucketPolicyActions(requestType, s.Action, log);
+        const resourceMatch = _checkBucketPolicyResources(request, s.Resource, log);
 
-        if (principalMatch && actionMatch && s.Effect === 'Deny') {
+        if (principalMatch && actionMatch && resourceMatch && s.Effect === 'Deny') {
```
Comment (miniscruff) [inline lib/api/apiUtils/authorization/permissionChecks.js:254]: The first 3 comparisons are the same for these two if statements so I would merge them.
```js
if ( a && b && c) {
    if ( d ) { blah }
}
```
( cant use the suggestion feature... )

Comment (miniscruff) [reply lib/api/apiUtils/authorization/permissionChecks.js:254]: Extra question: Why is this using a while loop and not for or something? Looks like a normal for ... array loop but it is instead a while + array thing. Not important just caught my eye.

Comment (naren-rajendran) [reply lib/api/apiUtils/authorization/permissionChecks.js:254]: Not sure why it was implemented this way!

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
Note to reviewers, I am still checking to see if more tests are needed!

**Review by miniscruff — APPROVED** [review]

**Review by alexanderchan-scality — APPROVED** [review]

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
@bert-e status

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
@bert-e status

**Pull request merged**